When a new release is made, bump the `CHANGELOG` version in the next release branch and the originating release branch

weierophinney opened the pull request:
| Q | A |
| - | - |
| New Feature | yes |
| RFC | #33 |
| BC Break | ? |

This patch adds a new interface, `BumpAndCommitChangelogVersion`, which describes classes that can update a changelog file to add a new version, and then commit and push the changes. I have included one implementation, `BumpAndCommitChangelogVersionViaKeepAChangelog` which uses utilities from phly/keep-a-changelog, as well as internal utilities, to do this.

I have added `BumpAndCommitChangelogVersion` as a dependency on `SwitchDefaultBranchToNextMinor`. The functionality is invoked only if a new release branch is made, and then immediately following that operation. At that time, the changelog is updated with a stub for the next **minor** version.

I have created a new command, `laminas:releases:bump-changelog`, implemented by `BumpChangelogForReleaseBranch`, which determines the release version and release branch, and uses those to call the composed `BumpAndCommitChangelogVersion`, this time bumping the changelog to the next **patch** version on the **originating** release branch.

Finally, I updated the sample workflow to call `laminas:releases:bump-changelog` as the final action in the workflow.  This approach guarantees that the newly-created `vNEXT` release branch does NOT contain a stub for the next patch release; only the originating branch will.

Fixes #33

weierophinney committed:
feat: functionality for bumping, commiting, and pushing new changelog version entries

This patch introduces the interface `BumpAndCommitChangelogVersion`,
describing an operation where a new changelog entry is injected into the
changelog file, if it exists; the new version is determined by the first
argument passed, which must be one of a set of known constant values
describing patch and minor release bumps.

This patch also introduces the `CheckoutBranch` interface, describing
how to checkout a specific branch of a repository. It has one
implementation, `CheckoutBranchViaConsole`.

The new class `BumpAndCommitChangelogVersionViaKeepAChangelog` composes
a `CheckoutBranch` instance, as well as `CommitFile`, `Push`, and
`LoggerInterface` instances, to assist in the work of updating the
changelog. If no changelog is detected in the branch, no work is done.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: when a new release branch is created, and a changelog file exists, bump the changelog file to the next minor release, commit, and push.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: add BumpChangelogForReleaseBranch command

Command determines the name of the release from the closed milestone,
and uses it to bump the changelog version, commit, and push back to the
associated release branch.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: bump changelog version on originating release branch on completion of other actions

By doing this as a separate command, we can delay adding the patch-level
changelog stub to the originating release branch until **after** we have
created the next release branch.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented on examples/.github/workflows/release-on-milestone-closed.yml:
```diff
@@ -45,3 +45,13 @@ jobs:
           "GIT_AUTHOR_NAME": ${{ secrets.GIT_AUTHOR_NAME }}
           "GIT_AUTHOR_EMAIL": ${{ secrets.GIT_AUTHOR_EMAIL }}
+
+      - name: "Bump Changelog Version On Originating Release Branch"
```
Should be done also in `.github`, not just in examples

weierophinney replied:
oops!

Ocramius commented on src/Changelog/BumpAndCommitChangelogVersion.php:
```diff
@@ -0,0 +1,30 @@
+
+interface BumpAndCommitChangelogVersion
+{
+    public const BUMP_MINOR = 'bumpMinorVersion';
```
Not a fan of these public constants to switch behaviour

Ocramius replied:
Hmm, unsure about this one: possibly good enough, since we enforce it at type level anyway 🤔

Ocramius commented on src/Changelog/BumpAndCommitChangelogVersion.php:
```diff
@@ -0,0 +1,30 @@
+    ];
+
+    /**
+     * @psalm-param value-of<self::KNOWN_BUMP_TYPES> $bumpType
```
Can be `self::BUMP_*` (actual type syntax), so we don't need to declare the array

Ocramius commented on src/Changelog/BumpAndCommitChangelogVersionViaKeepAChangelog.php:
```diff
@@ -0,0 +1,85 @@
+    ): void {
+        ($this->checkoutBranch)($repositoryDirectory, $sourceBranch->name());
+
+        $changelogFile = sprintf('%s/%s', $repositoryDirectory, self::CHANGELOG_FILE);
```
Not a problem, but sprintf for concatenating `/`? 😅

weierophinney replied:
I find it easier to read. :stuck_out_tongue:

Ocramius replied:
That's alright, just exotic 👍

Ocramius commented on src/Git/CheckoutBranch.php:
```diff
@@ -0,0 +1,17 @@
+{
+    /**
+     * @psalm-param non-empty-string $repositoryDirectory
+     * @psalm-param non-empty-string $branchName
```
We have a value type for `BranchName` in the repo: let's use that 👍

weierophinney replied:
I'd actually originally done that, but reverted because another class was taking just the string value. Updating again...

Ocramius commented on src/Git/CheckoutBranchViaConsole.php:
```diff
@@ -0,0 +1,18 @@
+        string $branchName
+    ): void {
+        (new Process(['git', 'checkout', $branchName], $repositoryDirectory))
+            ->mustRun();
```
I wonder if we should use `switch` these days

weierophinney replied:
Yes, we should. I'll update here, and elsewhere.

weierophinney replied:
Actually... will only update in the stuff relevant to this particular patch. I'll do a separate PR to update this everywhere later.

Ocramius commented on bin/console.php:
```diff
@@ -123,7 +133,15 @@ static function (int $errorCode, string $message = '', string $file = '', int $l
+            ),
+            $bumpChangelogVersion
+        ),
+        new BumpChangelogForReleaseBranch(
```
I like that this is a new command to be explicitly run after the release 👍

weierophinney committed:
qa: add laminas:automatic-releases:bump-changelog to repo workflow

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: simplify type definition for BumpAndCommitChangelogVersion `$bumpType` argument

Can use `{class}::TYPE_*` annotations instead of declaring an additional
constant.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: use BranchName and not a non-empty-string withig `CheckoutBranch`

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: use git switch instead of git checkout to switch branches

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: Adds entries to CHANGELOG for #40

- Details new laminas:automatic-releases:bump-changelog command
- Details changes to "switch default branch to next minor" command

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@Ocramius I've pushed changes addressing the feedback!

Ocramius reviewed (approved):
Excellent, thanks @weierophinney!
